File: world/habitat.py
"""
Habitat system - manages owned items and their placement in the duck's home.
"""
from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass, field
import json
from pathlib import Path

from world.shop import ShopItem, get_item, ItemCategory


# Fixed positions for items - ensures consistent placement
# Grid is 20x12, organized by category:
# - Toys: left side (x: 2-7)
# - Water features: center-right (x: 10-15)
# - Furniture: right side (x: 15-18)
# - Decorations/Plants: scattered around edges
ITEM_FIXED_POSITIONS: Dict[str, Tuple[int, int]] = {
    # === TOYS (left side) ===
    "toy_ball": (3, 3),
    "toy_blocks": (5, 2),
    "toy_trumpet": (2, 5),
    "toy_skateboard": (4, 7),
    "toy_piano": (6, 4),
    "toy_trampoline": (3, 9),
    "toy_slide": (5, 6),
    "toy_swing": (2, 8),
    "toy_seesaw": (6, 10),
    "toy_sandbox": (4, 10),
    "toy_boombox": (7, 3),
    "toy_car": (3, 6),
    "rubber_duck": (5, 5),
    "frisbee": (2, 2),
    "squeaky_toy": (6, 8),
    "trampoline": (4, 4),

    # === WATER FEATURES (center-right) ===
    "pool_kiddie": (11, 5),
    "pool_large": (12, 7),
    "fountain_small": (10, 3),
    "fountain_grand": (13, 4),
    "sprinkler": (11, 9),
    "hot_tub": (14, 6),
    "birdbath": (10, 8),
    "pond": (12, 2),
    "water_slide": (13, 9),
    "waterfall": (14, 3),
    "fountain_statue": (11, 2),
    "birdbath_garden": (10, 10),
    "fountain_decorative": (13, 2),

    # === FURNITURE (right side) ===
    "chair_wood": (16, 4),
    "chair_throne": (17, 3),
    "table_small": (15, 5),
    "bed_small": (18, 6),
    "bed_king": (17, 8),
    "couch": (16, 7),
    "hammock": (18, 4),
    "bookshelf": (15, 3),
    "lamp": (18, 2),
    "fireplace": (16, 2),
    "piano": (15, 9),

    # === PLANTS (scattered around edges) ===
    "flower_rose": (1, 1),
    "flower_tulip": (19, 1),
    "flower_sunflower": (1, 10),
    "tree_oak": (19, 10),
    "tree_pine": (1, 6),
    "cactus": (19, 6),
    "bamboo": (8, 1),
    "bonsai": (8, 10),

    # === DECORATIONS ===
    "rock_pile": (9, 2),
    "garden_gnome": (9, 9),
    "wind_chime": (8, 5),
    "bird_feeder": (9, 7),
}


@dataclass
class PlacedItem:
    """An item that has been placed in the habitat."""
    item_id: str
    x: int  # Position in habitat grid
    y: int
    last_interaction: float = 0  # Timestamp of last duck interaction
    # Animation state
    anim_offset_x: float = 0.0  # Temporary x offset for animation
    anim_offset_y: float = 0.0  # Temporary y offset for animation
    anim_start: float = 0.0  # When animation started
    is_animating: bool = False

    def to_dict(self) -> dict:
        return {
            "item_id": self.item_id,
            "x": self.x,
            "y": self.y,
            "last_interaction": self.last_interaction
            # Don't save animation state
        }

    @staticmethod
    def from_dict(data: dict) -> "PlacedItem":
        return PlacedItem(
            item_id=data.get("item_id", ""),
            x=data.get("x", 0),
            y=data.get("y", 0),
            last_interaction=data.get("last_interaction", 0)
        )

    def start_animation(self, anim_type: str = "bounce"):
        """Start an interaction animation."""
        import time
        self.is_animating = True
        self.anim_start = time.time()
        # Different animation types
        if anim_type == "bounce":
            self.anim_offset_y = -1  # Jump up
        elif anim_type == "shake":
            self.anim_offset_x = 1  # Shake right first
        elif anim_type == "roll":
            self.anim_offset_x = 2  # Roll to the side

    def update_animation(self) -> bool:
        """Update animation state. Returns True if still animating."""
        import time
        if not self.is_animating:
            return False

        elapsed = time.time() - self.anim_start
        duration = 0.5  # Animation duration in seconds

        if elapsed >= duration:
            # Animation complete
            self.is_animating = False
            self.anim_offset_x = 0.0
            self.anim_offset_y = 0.0
            return False

        # Decay animation offset back to 0
        progress = elapsed / duration
        decay = 1.0 - progress
        self.anim_offset_x *= decay * 0.9
        self.anim_offset_y *= decay * 0.9

        # Add bounce effect
        if abs(self.anim_offset_y) > 0.1:
            import math
            # Bounce effect
            bounce = math.sin(progress * math.pi * 3) * decay
            self.anim_offset_y = bounce * -1

        return True

    def get_display_position(self) -> Tuple[int, int]:
        """Get position with animation offset applied."""
        return (
            self.x + int(round(self.anim_offset_x)),
            self.y + int(round(self.anim_offset_y))
        )


class Habitat:
    """Manages the duck's habitat with owned and placed items."""
    
    def __init__(self):
        self.owned_items: List[str] = []  # Item IDs owned
        self.placed_items: List[PlacedItem] = []  # Items placed in habitat
        self.stored_items: List[str] = []  # Item IDs that are hidden/stored (not visible)
        self.equipped_cosmetics: Dict[str, str] = {}  # slot -> item_id
        self.currency: int = 100  # Starting currency
        
        # Habitat grid size (for placement)
        self.width = 20
        self.height = 12
    
    def add_currency(self, amount: int):
        """Add currency."""
        self.currency += amount
    
    def can_afford(self, cost: int) -> bool:
        """Check if player can afford an item."""
        return self.currency >= cost
    
    def purchase_item(self, item_id: str) -> bool:
        """Purchase an item if affordable."""
        item = get_item(item_id)
        if not item:
            return False
        
        if not self.can_afford(item.cost):
            return False
        
        if item_id in self.owned_items:
            return False  # Already owned
        
        self.currency -= item.cost
        self.owned_items.append(item_id)
        
        # Auto-equip cosmetics
        if item.category == ItemCategory.COSMETIC:
            self.equip_cosmetic(item_id)
        else:
            # Auto-place non-cosmetic items in the habitat
            self._auto_place_item(item_id)
        
        return True
    
    def _auto_place_item(self, item_id: str):
        """Place an item at its designated fixed position, or use deterministic fallback."""
        # First, check for a fixed position
        if item_id in ITEM_FIXED_POSITIONS:
            x, y = ITEM_FIXED_POSITIONS[item_id]
            if not self.get_item_at(x, y):
                self.place_item(item_id, x, y)
                return
            # Fixed position occupied, try nearby positions
            for dx, dy in [(1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (-1, -1)]:
                nx, ny = x + dx, y + dy
                if 1 <= nx < self.width - 1 and 1 <= ny < self.height - 1:
                    if not self.get_item_at(nx, ny):
                        self.place_item(item_id, nx, ny)
                        return

        # Fallback: hash-based deterministic placement (same item always tries same spot)
        hash_val = hash(item_id)
        base_x = (abs(hash_val) % 16) + 2  # Range 2-17
        base_y = (abs(hash_val >> 8) % 8) + 2  # Range 2-9

        # Avoid the center area where duck spawns
        if 7 <= base_x <= 12 and 4 <= base_y <= 8:
            base_x = (base_x + 8) % 16 + 2

        # Try the deterministic position first
        if not self.get_item_at(base_x, base_y):
            self.place_item(item_id, base_x, base_y)
            return

        # If occupied, spiral outward from the deterministic position
        for radius in range(1, max(self.width, self.height)):
            for dx in range(-radius, radius + 1):
                for dy in range(-radius, radius + 1):
                    if abs(dx) != radius and abs(dy) != radius:
                        continue  # Only check perimeter
                    nx, ny = base_x + dx, base_y + dy
                    if 1 <= nx < self.width - 1 and 1 <= ny < self.height - 1:
                        # Avoid center duck spawn area
                        if 7 <= nx <= 12 and 4 <= ny <= 8:
                            continue
                        if not self.get_item_at(nx, ny):
                            self.place_item(item_id, nx, ny)
                            return
    
    def owns_item(self, item_id: str) -> bool:
        """Check if player owns an item."""
        return item_id in self.owned_items
    
    def place_item(self, item_id: str, x: int, y: int) -> bool:
        """Place an owned item in the habitat."""
        if not self.owns_item(item_id):
            return False
        
        # Check if position is valid
        if x < 0 or x >= self.width or y < 0 or y >= self.height:
            return False
        
        # Check if something is already there
        if self.get_item_at(x, y):
            return False
        
        placed = PlacedItem(item_id=item_id, x=x, y=y)
        self.placed_items.append(placed)
        return True
    
    def remove_item_at(self, x: int, y: int) -> bool:
        """Remove an item from a position."""
        item = self.get_item_at(x, y)
        if item:
            self.placed_items.remove(item)
            return True
        return False
    
    def get_item_at(self, x: int, y: int) -> Optional[PlacedItem]:
        """Get the item at a position."""
        for item in self.placed_items:
            if item.x == x and item.y == y:
                return item
        return None

    def update_animations(self):
        """Update all item animations."""
        for item in self.placed_items:
            item.update_animation()

    def animate_item(self, placed_item: PlacedItem, anim_type: str = "bounce"):
        """Start an animation on a placed item."""
        placed_item.start_animation(anim_type)

    def get_items_near(self, x: int, y: int, radius: int = 3) -> List[PlacedItem]:
        """Get visible items near a position (excludes stored items)."""
        nearby = []
        for item in self.placed_items:
            # Skip stored (hidden) items
            if item.item_id in self.stored_items:
                continue
            dx = abs(item.x - x)
            dy = abs(item.y - y)
            if dx <= radius and dy <= radius:
                nearby.append(item)
        return nearby

    def get_placed_items_by_category(self, category: ItemCategory) -> List[PlacedItem]:
        """Get all visible placed items of a category (excludes stored items)."""
        result = []
        for placed in self.placed_items:
            # Skip stored (hidden) items
            if placed.item_id in self.stored_items:
                continue
            item = get_item(placed.item_id)
            if item and item.category == category:
                result.append(placed)
        return result
    
    def equip_cosmetic(self, item_id: str):
        """Equip a cosmetic item."""
        item = get_item(item_id)
        if not item or item.category != ItemCategory.COSMETIC:
            return
        
        # Determine slot based on item name
        slot = self._get_cosmetic_slot(item_id)
        self.equipped_cosmetics[slot] = item_id
    
    def unequip_cosmetic(self, slot: str):
        """Remove a cosmetic from a slot."""
        if slot in self.equipped_cosmetics:
            del self.equipped_cosmetics[slot]
    
    def _get_cosmetic_slot(self, item_id: str) -> str:
        """Determine which slot a cosmetic goes in."""
        # Hats and head items
        head_items = ["hat_", "cap_", "beanie", "crown", "helmet", "wizard", "pirate", "viking", 
                      "party_hat", "flower_crown", "tiara", "graduation", "nurse", "pilot", 
                      "detective", "cat_ears", "bunny_ears", "antenna", "propeller", "jester"]
        for h in head_items:
            if h in item_id:
                return "head"
        
        # Glasses and face items  
        if "glasses_" in item_id or "sunglasses" in item_id or "monocle" in item_id:
            return "eyes"
        
        # Neck accessories
        if "bowtie" in item_id or "bow_tie" in item_id or "scarf_" in item_id or "bandana" in item_id:
            return "neck"
        
        # Back items
        if "cape" in item_id or "wings_" in item_id or "backpack" in item_id:
            return "back"
        
        # Above head (floating)
        if "halo" in item_id or "devil_horns" in item_id or "headphones" in item_id:
            return "above"
        
        return "head"  # Default to head slot
    
    def get_nearby_items(self, x: int, y: int, radius: int = 2) -> List[PlacedItem]:
        """Get visible items near a position for duck interaction (excludes stored items)."""
        nearby = []
        for item in self.placed_items:
            # Skip stored (hidden) items
            if item.item_id in self.stored_items:
                continue
            dx = abs(item.x - x)
            dy = abs(item.y - y)
            if dx <= radius and dy <= radius:
                nearby.append(item)
        return nearby
    
    def mark_interaction(self, placed_item: PlacedItem, timestamp: float):
        """Mark that the duck interacted with an item."""
        placed_item.last_interaction = timestamp
    
    def save(self, filepath: Path):
        """Save habitat data to file."""
        data = {
            "owned_items": self.owned_items,
            "placed_items": [item.to_dict() for item in self.placed_items],
            "equipped_cosmetics": self.equipped_cosmetics,
            "currency": self.currency
        }
        filepath.parent.mkdir(parents=True, exist_ok=True)
        with open(filepath, 'w') as f:
            json.dump(data, f, indent=2)
    
    def load(self, filepath: Path):
        """Load habitat data from file."""
        if not filepath.exists():
            return
        
        try:
            with open(filepath, 'r') as f:
                data = json.load(f)
            
            self.owned_items = data.get("owned_items", [])
            self.placed_items = [PlacedItem.from_dict(item) for item in data.get("placed_items", [])]
            self.equipped_cosmetics = data.get("equipped_cosmetics", {})
            self.currency = data.get("currency", 100)
        except Exception as e:
            print(f"Error loading habitat: {e}")
    
    def to_dict(self) -> dict:
        """Convert habitat to dictionary for saving."""
        return {
            "owned_items": self.owned_items,
            "placed_items": [item.to_dict() for item in self.placed_items],
            "stored_items": self.stored_items,
            "equipped_cosmetics": self.equipped_cosmetics,
            "currency": self.currency
        }
    
    def from_dict(self, data: dict):
        """Load habitat from dictionary."""
        self.owned_items = data.get("owned_items", [])
        self.placed_items = [PlacedItem.from_dict(item) for item in data.get("placed_items", [])]
        self.stored_items = data.get("stored_items", [])
        self.equipped_cosmetics = data.get("equipped_cosmetics", {})
        self.currency = data.get("currency", 100)
    
    def get_stats(self) -> Dict[str, int]:
        """Get habitat statistics."""
        return {
            "owned_items": len(self.owned_items),
            "placed_items": len(self.placed_items),
            "stored_items": len(self.stored_items),
            "currency": self.currency,
            "cosmetics_equipped": len(self.equipped_cosmetics)
        }
    
    def is_item_stored(self, item_id: str) -> bool:
        """Check if an owned item is currently stored (hidden)."""
        return item_id in self.stored_items
    
    def is_item_placed(self, item_id: str) -> bool:
        """Check if an owned item is currently placed (visible)."""
        return any(p.item_id == item_id for p in self.placed_items)
    
    def toggle_item_visibility(self, item_id: str) -> str:
        """Toggle an owned item between placed and stored states.
        
        Returns:
            'placed' - item is now visible in habitat
            'stored' - item is now hidden
            'not_owned' - item is not owned
            'cosmetic' - cosmetics can't be toggled this way
        """
        if item_id not in self.owned_items:
            return 'not_owned'
        
        # Cosmetics are handled separately via equip/unequip
        item = get_item(item_id)
        if item and item.category == ItemCategory.COSMETIC:
            return 'cosmetic'
        
        if item_id in self.stored_items:
            # Currently stored -> place it
            self.stored_items.remove(item_id)
            self._auto_place_item(item_id)
            return 'placed'
        else:
            # Currently placed -> store it
            # Remove from placed_items
            self.placed_items = [p for p in self.placed_items if p.item_id != item_id]
            self.stored_items.append(item_id)
            return 'stored'
    
    def store_item(self, item_id: str) -> bool:
        """Store (hide) an item from the habitat. Returns True if successful."""
        if item_id not in self.owned_items:
            return False
        
        item = get_item(item_id)
        if item and item.category == ItemCategory.COSMETIC:
            return False
        
        if item_id in self.stored_items:
            return True  # Already stored
        
        # Remove from placed_items
        self.placed_items = [p for p in self.placed_items if p.item_id != item_id]
        self.stored_items.append(item_id)
        return True
    
    def show_item(self, item_id: str) -> bool:
        """Show (place) an item in the habitat. Returns True if successful."""
        if item_id not in self.owned_items:
            return False
        
        item = get_item(item_id)
        if item and item.category == ItemCategory.COSMETIC:
            return False
        
        if item_id in self.stored_items:
            self.stored_items.remove(item_id)
        
        # Only place if not already placed
        if not any(p.item_id == item_id for p in self.placed_items):
            self._auto_place_item(item_id)
        
        return True
    
    def get_visible_placed_items(self) -> List[PlacedItem]:
        """Get only the placed items that are visible (not stored)."""
        return [p for p in self.placed_items if p.item_id not in self.stored_items]

File: StupidDuckCS/src/StupidDuck.World/Habitat.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;

namespace StupidDuck.World;

/// <summary>
/// An item that has been placed in the habitat.
/// </summary>
public class PlacedItem
{
    public string ItemId { get; set; } = "";
    public int X { get; set; }
    public int Y { get; set; }
    public double LastInteraction { get; set; }
    
    // Animation state (not saved)
    public double AnimOffsetX { get; set; }
    public double AnimOffsetY { get; set; }
    public double AnimStart { get; set; }
    public bool IsAnimating { get; set; }
    
    /// <summary>
    /// Start an interaction animation.
    /// </summary>
    public void StartAnimation(string animType = "bounce")
    {
        IsAnimating = true;
        AnimStart = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0;
        
        switch (animType)
        {
            case "bounce":
                AnimOffsetY = -1;
                break;
            case "shake":
                AnimOffsetX = 1;
                break;
            case "roll":
                AnimOffsetX = 2;
                break;
        }
    }
    
    /// <summary>
    /// Update animation state. Returns True if still animating.
    /// </summary>
    public bool UpdateAnimation()
    {
        if (!IsAnimating)
            return false;
        
        double now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0;
        double elapsed = now - AnimStart;
        double duration = 0.5;
        
        if (elapsed >= duration)
        {
            IsAnimating = false;
            AnimOffsetX = 0;
            AnimOffsetY = 0;
            return false;
        }
        
        double progress = elapsed / duration;
        double decay = 1.0 - progress;
        AnimOffsetX *= decay * 0.9;
        AnimOffsetY *= decay * 0.9;
        
        // Add bounce effect
        if (Math.Abs(AnimOffsetY) > 0.1)
        {
            double bounce = Math.Sin(progress * Math.PI * 3) * decay;
            AnimOffsetY = bounce * -1;
        }
        
        return true;
    }
    
    /// <summary>
    /// Get position with animation offset applied.
    /// </summary>
    public (int x, int y) GetDisplayPosition()
    {
        return (X + (int)Math.Round(AnimOffsetX), Y + (int)Math.Round(AnimOffsetY));
    }
    
    public Dictionary<string, object?> ToSaveData()
    {
        return new Dictionary<string, object?>
        {
            ["item_id"] = ItemId,
            ["x"] = X,
            ["y"] = Y,
            ["last_interaction"] = LastInteraction
        };
    }
    
    public static PlacedItem FromSaveData(JsonElement data)
    {
        return new PlacedItem
        {
            ItemId = data.TryGetProperty("item_id", out var idEl) ? idEl.GetString() ?? "" : "",
            X = data.TryGetProperty("x", out var xEl) ? xEl.GetInt32() : 0,
            Y = data.TryGetProperty("y", out var yEl) ? yEl.GetInt32() : 0,
            LastInteraction = data.TryGetProperty("last_interaction", out var liEl) ? liEl.GetDouble() : 0
        };
    }
}

/// <summary>
/// Habitat system - manages owned items and their placement in the duck's home.
/// </summary>
public class Habitat
{
    private static readonly Random _random = new();
    
    public List<string> OwnedItems { get; private set; } = new();
    public List<PlacedItem> PlacedItems { get; private set; } = new();
    public Dictionary<string, string> EquippedCosmetics { get; private set; } = new();
    public int Currency { get; set; } = 100;
    
    // Habitat grid size
    public int Width { get; } = 20;
    public int Height { get; } = 12;
    
    /// <summary>
    /// Add currency.
    /// </summary>
    public void AddCurrency(int amount)
    {
        Currency += amount;
    }
    
    /// <summary>
    /// Check if player can afford an item.
    /// </summary>
    public bool CanAfford(int cost)
    {
        return Currency >= cost;
    }
    
    /// <summary>
    /// Purchase an item if affordable.
    /// </summary>
    public bool PurchaseItem(string itemId, int cost, string category)
    {
        if (!CanAfford(cost))
            return false;
        
        if (OwnedItems.Contains(itemId))
            return false;
        
        Currency -= cost;
        OwnedItems.Add(itemId);
        
        // Auto-equip cosmetics
        if (category == "cosmetic")
        {
            EquipCosmetic(itemId);
        }
        else
        {
            // Auto-place non-cosmetic items in the habitat
            AutoPlaceItem(itemId);
        }
        
        return true;
    }
    
    /// <summary>
    /// Find a random empty spot and place an item.
    /// </summary>
    private void AutoPlaceItem(string itemId)
    {
        int maxAttempts = 50;
        for (int i = 0; i < maxAttempts; i++)
        {
            int x = _random.Next(1, Width - 1);
            int y = _random.Next(1, Height - 1);
            
            // Avoid the center area where duck typically is
            if (x >= 7 && x <= 12 && y >= 4 && y <= 8)
                continue;
            
            if (GetItemAt(x, y) == null)
            {
                PlaceItem(itemId, x, y);
                return;
            }
        }
        
        // If all attempts failed, place at first available spot
        for (int y = 1; y < Height - 1; y++)
        {
            for (int x = 1; x < Width - 1; x++)
            {
                if (GetItemAt(x, y) == null)
                {
                    PlaceItem(itemId, x, y);
                    return;
                }
            }
        }
    }
    
    /// <summary>
    /// Check if player owns an item.
    /// </summary>
    public bool OwnsItem(string itemId)
    {
        return OwnedItems.Contains(itemId);
    }
    
    /// <summary>
    /// Place an owned item in the habitat.
    /// </summary>
    public bool PlaceItem(string itemId, int x, int y)
    {
        if (!OwnsItem(itemId))
            return false;
        
        if (x < 0 || x >= Width || y < 0 || y >= Height)
            return false;
        
        if (GetItemAt(x, y) != null)
            return false;
        
        PlacedItems.Add(new PlacedItem { ItemId = itemId, X = x, Y = y });
        return true;
    }
    
    /// <summary>
    /// Remove an item from a position.
    /// </summary>
    public bool RemoveItemAt(int x, int y)
    {
        var item = GetItemAt(x, y);
        if (item != null)
        {
            PlacedItems.Remove(item);
            return true;
        }
        return false;
    }
    
    /// <summary>
    /// Get the item at a position.
    /// </summary>
    public PlacedItem? GetItemAt(int x, int y)
    {
        return PlacedItems.FirstOrDefault(i => i.X == x && i.Y == y);
    }
    
    /// <summary>
    /// Update all item animations.
    /// </summary>
    public void UpdateAnimations()
    {
        foreach (var item in PlacedItems)
        {
            item.UpdateAnimation();
        }
    }
    
    /// <summary>
    /// Start an animation on a placed item.
    /// </summary>
    public void AnimateItem(PlacedItem placedItem, string animType = "bounce")
    {
        placedItem.StartAnimation(animType);
    }
    
    /// <summary>
    /// Get items near a position.
    /// </summary>
    public List<PlacedItem> GetItemsNear(int x, int y, int radius = 3)
    {
        var nearby = new List<PlacedItem>();
        foreach (var item in PlacedItems)
        {
            int dx = Math.Abs(item.X - x);
            int dy = Math.Abs(item.Y - y);
            if (dx <= radius && dy <= radius)
                nearby.Add(item);
        }
        return nearby;
    }
    
    /// <summary>
    /// Get all placed items of a category.
    /// </summary>
    public List<PlacedItem> GetPlacedItemsByCategory(string category)
    {
        // Note: Requires ShopSystem integration to look up item categories
        // For now, returns all items
        return PlacedItems.ToList();
    }
    
    /// <summary>
    /// Equip a cosmetic item.
    /// </summary>
    public void EquipCosmetic(string itemId)
    {
        string slot = GetCosmeticSlot(itemId);
        EquippedCosmetics[slot] = itemId;
    }
    
    /// <summary>
    /// Remove a cosmetic from a slot.
    /// </summary>
    public void UnequipCosmetic(string slot)
    {
        EquippedCosmetics.Remove(slot);
    }
    
    /// <summary>
    /// Determine which slot a cosmetic goes in.
    /// </summary>
    private string GetCosmeticSlot(string itemId)
    {
        // Hats and head items
        string[] headItems = { "hat_", "cap_", "beanie", "crown", "helmet", "wizard", "pirate", "viking",
            "party_hat", "flower_crown", "tiara", "graduation", "nurse", "pilot",
            "detective", "cat_ears", "bunny_ears", "antenna", "propeller", "jester" };
        
        foreach (var h in headItems)
        {
            if (itemId.Contains(h))
                return "head";
        }
        
        // Glasses and face items
        if (itemId.Contains("glasses_") || itemId.Contains("sunglasses") || itemId.Contains("monocle"))
            return "eyes";
        
        // Neck accessories
        if (itemId.Contains("bowtie") || itemId.Contains("bow_tie") || itemId.Contains("scarf_") || itemId.Contains("bandana"))
            return "neck";
        
        // Back items
        if (itemId.Contains("cape") || itemId.Contains("wings_") || itemId.Contains("backpack"))
            return "back";
        
        // Above head (floating)
        if (itemId.Contains("halo") || itemId.Contains("devil_horns") || itemId.Contains("headphones"))
            return "above";
        
        return "head"; // Default to head slot
    }
    
    /// <summary>
    /// Get items near a position (for duck interaction).
    /// </summary>
    public List<PlacedItem> GetNearbyItems(int x, int y, int radius = 2)
    {
        return GetItemsNear(x, y, radius);
    }
    
    /// <summary>
    /// Mark that the duck interacted with an item.
    /// </summary>
    public void MarkInteraction(PlacedItem placedItem, double timestamp)
    {
        placedItem.LastInteraction = timestamp;
    }
    
    /// <summary>
    /// Get habitat statistics.
    /// </summary>
    public Dictionary<string, int> GetStats()
    {
        return new Dictionary<string, int>
        {
            ["owned_items"] = OwnedItems.Count,
            ["placed_items"] = PlacedItems.Count,
            ["currency"] = Currency,
            ["cosmetics_equipped"] = EquippedCosmetics.Count
        };
    }
    
    /// <summary>
    /// Render habitat display.
    /// </summary>
    public List<string> RenderHabitat()
    {
        var lines = new List<string>();
        
        // Create grid
        var grid = new char[Height, Width];
        for (int y = 0; y < Height; y++)
        {
            for (int x = 0; x < Width; x++)
            {
                if (y == 0 || y == Height - 1)
                    grid[y, x] = '═';
                else if (x == 0 || x == Width - 1)
                    grid[y, x] = '║';
                else
                    grid[y, x] = ' ';
            }
        }
        
        // Place items
        foreach (var item in PlacedItems)
        {
            var (dx, dy) = item.GetDisplayPosition();
            if (dx >= 0 && dx < Width && dy >= 0 && dy < Height)
            {
                grid[dy, dx] = '▪';
            }
        }
        
        // Corners
        grid[0, 0] = '╔';
        grid[0, Width - 1] = '╗';
        grid[Height - 1, 0] = '╚';
        grid[Height - 1, Width - 1] = '╝';
        
        // Convert to strings
        for (int y = 0; y < Height; y++)
        {
            var row = new char[Width];
            for (int x = 0; x < Width; x++)
                row[x] = grid[y, x];
            lines.Add(new string(row));
        }
        
        return lines;
    }
    
    /// <summary>
    /// Convert to save data.
    /// </summary>
    public Dictionary<string, object?> ToSaveData()
    {
        return new Dictionary<string, object?>
        {
            ["owned_items"] = OwnedItems,
            ["placed_items"] = PlacedItems.Select(i => i.ToSaveData()).ToList(),
            ["equipped_cosmetics"] = EquippedCosmetics,
            ["currency"] = Currency
        };
    }
    
    /// <summary>
    /// Load from save data.
    /// </summary>
    public static Habitat FromSaveData(Dictionary<string, JsonElement> data)
    {
        var habitat = new Habitat();
        
        if (data.TryGetValue("owned_items", out var ownedEl) && ownedEl.ValueKind == JsonValueKind.Array)
            habitat.OwnedItems = ownedEl.EnumerateArray().Select(e => e.GetString() ?? "").ToList();
        
        if (data.TryGetValue("placed_items", out var placedEl) && placedEl.ValueKind == JsonValueKind.Array)
            habitat.PlacedItems = placedEl.EnumerateArray().Select(PlacedItem.FromSaveData).ToList();
        
        if (data.TryGetValue("equipped_cosmetics", out var equippedEl) && equippedEl.ValueKind == JsonValueKind.Object)
        {
            foreach (var prop in equippedEl.EnumerateObject())
            {
                habitat.EquippedCosmetics[prop.Name] = prop.Value.GetString() ?? "";
            }
        }
        
        if (data.TryGetValue("currency", out var currEl))
            habitat.Currency = currEl.GetInt32();
        
        return habitat;
    }
    
    /// <summary>
    /// Load from save data dictionary.
    /// </summary>
    public void LoadFromData(Dictionary<string, JsonElement> data)
    {
        var loaded = FromSaveData(data);
        OwnedItems = loaded.OwnedItems;
        PlacedItems = loaded.PlacedItems;
        EquippedCosmetics = loaded.EquippedCosmetics;
        Currency = loaded.Currency;
    }
}
